atp/client.go

```diff
@@ -465,28 +520,34 @@ func (c *client) prepareResultChannels(
 	return nil
 }
 
-// getResultV2 works with the channels that communicate with the RuntimeMessage loop.
+// getResultV2 communicates with the RuntimeMessage loop to get the .
 func (c *client) getResultV2(
 	stepData schema.Input,
 ) ExecutionResult {
 	c.mutex.Lock()
-	resultChannel, found := c.runningStepResultChannels[stepData.RunID]
-	c.mutex.Unlock()
+	resultEntry, found := c.runningStepResultEntries[stepData.RunID]
+	c.logger.Debugf("Got result entry for run ID %q", stepData.RunID)
 	if !found {
 		return NewErrorExecutionResult(
-			fmt.Errorf("could not find result channel for step with run ID '%s'",
-				stepData.RunID),
+			fmt.Errorf("could not find result entry for step with run ID '%s'. Existing entries: %v",
+				stepData.RunID, c.runningStepResultEntries),
 		)
 	}
-	// Wait for the result
-	result, received := <-resultChannel
-	if !received {
+	if resultEntry.result == nil {
+		// Wait for the result
+		resultEntry.condition.Wait()
+	}
+	if resultEntry.result == nil {
 		return NewErrorExecutionResult(
-			fmt.Errorf("did not receive result from results channel in ATP client for step with run ID '%s'",
+			fmt.Errorf("did not receive result from results entry in ATP client for step with run ID '%s'",
 				stepData.RunID),
 		)
 	}
-	return result
+	defer c.mutex.Unlock()
+	// This needs to be done after receiving the value, or else the sender will
+	// not be able to get the entry.
+	delete(c.runningStepResultEntries, stepData.RunID)
```

Comment:
I don't understand what this comment is saying.  (Perhaps start by replacing "This" with what the pronoun is referring to!)  How does deleting the entry enable the sender to get it??

Reply:
I redid it.